build_runner/lib/src/generate/build_impl.dart

```diff
@@ -222,53 +233,103 @@ class BuildImpl {
-          'Outputs should be known statically. Missing '
-          '${builderOutputs.where((o) => !_assetGraph.contains(o)).toList()}');
-      // Skip the build step if none of the outputs need updating.
-      var skipBuild = !builderOutputs.any((output) =>
-          (_assetGraph.get(output) as GeneratedAssetNode).needsUpdate);
-      if (skipBuild) return;
+
+      if (!_buildShouldRun(builderOutputs)) return;
+
       var wrappedReader = new SinglePhaseReader(
           _reader, _assetGraph, phaseNumber, input.package);
       var wrappedWriter = new AssetWriterSpy(_writer);
       await runBuilder(
           builder, [input], wrappedReader, wrappedWriter, _resolvers,
           resourceManager: resourceManager);
 
-      // Mark all outputs as no longer needing an update, and mark `wasOutput`
-      // as `false` for now (this will get reset to true later one).
-      //
-      // Also tracks all the globs that were used during this build.
-      for (var output in builderOutputs) {
-        (_assetGraph.get(output) as GeneratedAssetNode)
-          ..needsUpdate = false
-          ..wasOutput = false
-          ..globs = wrappedReader.globsRan.toSet();
-      }
-
-      // Update the asset graph based on the dependencies discovered.
-      for (var dependency in wrappedReader.assetsRead) {
-        var dependencyNode = _assetGraph.get(dependency);
-        assert(dependencyNode != null, 'Asset Graph is missing $dependency');
-        // We care about all builderOutputs, not just real outputs. Updates
-        // to dependencies may cause a file to be output which wasn't before.
-        dependencyNode.outputs.addAll(builderOutputs);
-      }
+      // Reset the state for all the `builderOutputs` nodes based on what was
+      // read and written.
+      _setOutputsState(builderOutputs, wrappedReader, wrappedWriter);
 
-      // Yield the outputs.
-      for (var output in wrappedWriter.assetsWritten) {
-        (_assetGraph.get(output) as GeneratedAssetNode).wasOutput = true;
-        outputs.add(output);
-      }
+      outputs.addAll(wrappedWriter.assetsWritten);
     }
 
     await Future.wait(primaryInputs.map(runForInput));
 
     return outputs;
   }
 
+  /// Runs the [PackageBuilder] [builder] and returns only the outputs
+  /// that were newly created.
+  ///
+  /// Does not return outputs that didn't need to be re-ran or were declared
+  /// but not output.
+  Future<Iterable<AssetId>> _runPackageBuilder(int phaseNumber, String package,
+      PackageBuilder builder, ResourceManager resourceManager) async {
+    var builderOutputs = outputIdsForBuilder(builder, package);
+
+    if (!_buildShouldRun(builderOutputs)) return <AssetId>[];
+
+    var wrappedReader =
+        new SinglePhaseReader(_reader, _assetGraph, phaseNumber, package);
+    var wrappedWriter = new AssetWriterSpy(_writer);
+
+    var logger = new Logger('runBuilder');
+    var buildStep = new BuildStepImpl(null, builderOutputs, wrappedReader,
```

Comment:
we'll have to be more careful about what we consider breaking changes around this.

Reply:
Ya, I had to add two imports to `src` of package:build :(.